crates/vm/src/system/memory/paged_vec.rs

```diff
@@ -0,0 +1,368 @@
+use std::ops::Range;
+
+use serde::{Deserialize, Serialize};
+
+use crate::arch::MemoryConfig;
+
+/// (address_space, pointer)
+pub(crate) type Address = (u32, u32);
+pub(crate) const PAGE_SIZE: usize = 1 << 12;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub(crate) struct PagedVec<T, const PAGE_SIZE: usize> {
+    pages: Vec<Option<Vec<T>>>,
+}
+
+impl<T: Default + Clone, const PAGE_SIZE: usize> PagedVec<T, PAGE_SIZE> {
+    pub fn new(num_pages: usize) -> Self {
+        Self {
+            pages: vec![None; num_pages],
+        }
+    }
+
+    pub fn get(&self, index: usize) -> Option<&T> {
+        let page_idx = index / PAGE_SIZE;
+        self.pages[page_idx]
+            .as_ref()
+            .map(|page| &page[index % PAGE_SIZE])
+    }
+
+    pub fn get_mut(&mut self, index: usize) -> Option<&mut T> {
+        let page_idx = index / PAGE_SIZE;
+        self.pages[page_idx]
+            .as_mut()
+            .map(|page| &mut page[index % PAGE_SIZE])
+    }
+
+    pub fn set(&mut self, index: usize, value: T) -> Option<T> {
+        let page_idx = index / PAGE_SIZE;
+        if let Some(page) = self.pages[page_idx].as_mut() {
+            Some(std::mem::replace(&mut page[index % PAGE_SIZE], value))
+        } else {
+            let page = self.pages[page_idx].get_or_insert_with(|| vec![T::default(); PAGE_SIZE]);
+            page[index % PAGE_SIZE] = value;
+            None
+        }
+    }
+
+    pub fn get_range(&self, range: Range<usize>) -> Vec<T> {
+        let start_page_idx = range.start / PAGE_SIZE;
+        let end_page_idx = range.end / PAGE_SIZE;
+
+        if start_page_idx == end_page_idx {
+            if let Some(start_page) = &self.pages[start_page_idx] {
+                let i = range.start % PAGE_SIZE;
+                start_page[i..i + range.len()].to_vec()
+            } else {
+                vec![T::default(); range.len()]
+            }
+        } else {
+            // TODO: This can be more efficient by copying from two slices (but most queries should
+            // not be cross-page).
+            range
+                .map(|i| self.get(i).cloned().unwrap_or_default())
+                .collect()
+        }
+    }
+
+    pub fn set_range<'a>(
+        &mut self,
+        range: Range<usize>,
+        values: impl IntoIterator<Item = &'a T>,
+    ) -> Vec<T>
+    where
+        T: 'a,
+    {
+        let start_page_idx = range.start / PAGE_SIZE;
+        let end_page_idx = range.end / PAGE_SIZE;
+
+        if start_page_idx == end_page_idx {
+            let page =
+                self.pages[start_page_idx].get_or_insert_with(|| vec![T::default(); PAGE_SIZE]);
```

Comment:
Since we're going to copy the whole thing, we don't need to spend time initializing to default. We should optimize this.

Reply:
oh sorry, this is initializing the entire page, so we don't copy the whole thing.